### PR TITLE
BOLEClassifier + DriftRetrainingClassifier

### PR DESCRIPTION
**Pull request by danielnowakassis**

### CODE REVIEW

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:238]: You can do `self[pos][1].learn_one`, which avoids `models = list(enumerate(self))`

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:258]:
```suggestion
                    model_weight = math.log(beta_inv)
```

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:270]: I really like this class. In fact, I think it should be a class of its own that would live in the `drift` module. It's really useful I would call it something more generic, such as `RetrainOnDriftClassifier`

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:283]:
```suggestion
    def __init__(self, model: base.Classifier, drift_detector: DriftDetector = None):
```

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:284]: One thing you can do is only store the `model` attribute. Then, you can call `model.clone()` to get reinitialize the model, without having to store `base_class`

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:300]: It's preferable to do 

```py
if y_pred is None:
    return
```

and then continue. It's more readable that way

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:305]: You can use `.clone()` instead of `copy.deepcopy`

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:260]:
```suggestion
            if model_weight:
```

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:10]: Avoid using absolute imports. It's more Pythonic to import `base` and `drift` and then access the classes you want.

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:185]:
```suggestion

```

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:305]: I think you can do `self.bkg_model = copy.deepcopy(self.base_class)` (or `self.base_class.clone()`) in the `__init__` (on line 286) and then delete this `if`. Then you can call `self.bkg_model = copy.deepcopy(self.base_class)` (or `self.base_class.clone()`) at line 310

Comment (MaxHalford) [reply river/ensemble/boosting.py:305]: But maybe I'm wrong! I'll let you check :)

**Review by MaxHalford — CHANGES_REQUESTED** [review]
Hey! I made some comments. Overall I'm very pleased with the simplicity of the method, as well as the model's performance (apparently).

I really like the `BOLEBaseModel` class, and I think it should be a class of its own. People will want to use it.

There are some code style issues. I've commented most of them. Don't feel negative about them, we're just a bit maniac in our codebase :). Also, if it helps I can make a pull request on your pull request to suggest some changes.

Cheers!

File: river/ensemble/boosting.py
Comment (MaxHalford) [inline river/ensemble/boosting.py:188]: Would this model work with other drift detectors? In theory? Does it have to just for `ADWIN`?

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: resolved in snake_case + range() commit

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: resolved in snake_case + range() commit. acc in learn_one will be replaced with [0] * 9. If you have a better suggestion, please comment :)

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: the object is not subscriptable. Solved in last numpy removes and self.models[ind] commit. River has a self.models that returns the models.

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in last numpy removals and self.models[ind] commit.

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py:267]: It´s better to do this in another pull-request.

Comment (danielnowakassis) [reply river/ensemble/boosting.py:267]: I will substitute the changes suggested and then take BOLEBaseModel out

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py:281]: For me it is not clear what you intend. Can you modify?

Comment (danielnowakassis) [reply river/ensemble/boosting.py:281]: For me it is not clear what you want. Can you modify it?

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py:280]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py:302]: I use None to avoid training without having a model (and maybe avoid some bugs).

Comment (danielnowakassis) [reply river/ensemble/boosting.py:302]: What's the difference in this context?

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: It's hard to explain, but i will do my best. Tried to solve in BaseModel + other changes

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: Solved in BaseModel + other changes

File: river/ensemble/boosting.py
Comment (danielnowakassis) [reply river/ensemble/boosting.py]: * self.n_models *

**Comment by MaxHalford** [comment]
Looking good @danielnowakassis! Could you give me push access to your fork? I'll push a few changes if that's ok :)

**Comment by MaxHalford** [comment]
Ok @danielnowakassis I pushed a fair number of changes. I'll let you take a look. What do you think?

**Comment by danielnowakassis** [comment]
I agree with all changes @MaxHalford. We have done some good work :) You can merge it.

**Comment by MaxHalford** [comment]
Agreed! Before merging though, we still need to fix the tests. On my end I'll look into test for `DriftRetrainingClassifier`. Can you check for `BOLEClassifier`?

**Comment by MaxHalford** [comment]
Good job @danielnowakassis!

**Comment by danielnowakassis** [comment]
Thanks Max. We have done some good work 🙂

**Pull request merged**
